### ISSUE TITLE
Errors in the output of the convolutional layer when batch sizes different

### ISSUE THREAD

**Issue by NGC13009** [pytorch/pytorch#134908]
### 🐛 Describe the bug

In theory, the output of any convolutional layer should be independent of the batch_size. For example, when I have a convolutional layer, once the input features [ci, h, w] are determined, the values of the output features [co, h, w] should also be determined and not related to the number of features I input (on batch dim), nor to the platform on which I run the code (CUDA or CPU, etc.)

However, through the following code, it can be observed that when running on the CPU platform, there is an error of approximately 1e-6 in the results obtained by batch feeding and individual sample feeding.

```python
import torch
from torch.nn import Conv2d

torch.manual_seed(13009)        # Fixed random number seed

ci = 4              # channel in  : This problem only occurs when the number of input and output channels is greater than 1
co = 3              # channel out : This problem only occurs when the number of input and output channels is greater than 1
b = 2               # batch size, if change this, you need to synchronize and modify the following code.
h = 8               # feature's high
w = 6               # feature's wide
k = 3               # kernel size
device = "cpu"      # problem only occurs when use cpu

c2d = Conv2d(ci, co, k, 1, 0, 1, device=device)     # a simple net

# # If you need to manually set the convolution core weights
# w1 = torch.randn(co, ci, k,k)
# b1 = torch.randn(co)
# c2d.weight.data = w1
# c2d.bias.data = b1

for i in range(10):     # Multiple tests
    x = torch.randn(2, ci, h, w, device=device)
    y1 = c2d(x)                                                             # One batch in
    y2 = torch.cat([c2d(x[0:1, :, :, :]), c2d(x[1:2, :, :, :])], dim=0)     # One by one
    diff = torch.sum(torch.abs(y2 - y1), dim=[1, 2, 3])                     # the different sum of feature per batch
    print(diff)
```

you will get the result:

```text
tensor([3.5465e-06, 5.0913e-06], grad_fn=<SumBackward1>)
tensor([4.6017e-06, 4.5355e-06], grad_fn=<SumBackward1>)
tensor([3.7085e-06, 3.9315e-06], grad_fn=<SumBackward1>)
tensor([5.0266e-06, 4.3502e-06], grad_fn=<SumBackward1>)
tensor([4.0308e-06, 3.9656e-06], grad_fn=<SumBackward1>)
tensor([3.4748e-06, 3.9535e-06], grad_fn=<SumBackward1>)
tensor([4.3921e-06, 2.4401e-06], grad_fn=<SumBackward1>)
tensor([4.0866e-06, 4.1882e-06], grad_fn=<SumBackward1>)
tensor([3.3006e-06, 4.2040e-06], grad_fn=<SumBackward1>)
tensor([3.6065e-06, 4.2403e-06], grad_fn=<SumBackward1>)
```

But there shouldn't be this error, theoretically it should all be 0. Using CUDA will get the correct result
```text
tensor([0., 0.], device='cuda:0', grad_fn=<SumBackward1>)
tensor([0., 0.], device='cuda:0', grad_fn=<SumBackward1>)
tensor([0., 0.], device='cuda:0', grad_fn=<SumBackward1>)
tensor([0., 0.], device='cuda:0', grad_fn=<SumBackward1>)
tensor([0., 0.], device='cuda:0', grad_fn=<SumBackward1>)
tensor([0., 0.], device='cuda:0', grad_fn=<SumBackward1>)
tensor([0., 0.], device='cuda:0', grad_fn=<SumBackward1>)
tensor([0., 0.], device='cuda:0', grad_fn=<SumBackward1>)
tensor([0., 0.], device='cuda:0', grad_fn=<SumBackward1>)
tensor([0., 0.], device='cuda:0', grad_fn=<SumBackward1>)
```

Batch size can cause errors (Approximately 1e-6 errors) in output features, only occurs when:
1. use cpu ( i try many cpus always occurs this problem)
2. number of input and output channels is both greater than 1

conv1d also have this error.

### Versions

Collecting environment information...
PyTorch version: 2.0.1+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 11 pro
GCC version: (Rev4, Built by MSYS2 project) 13.1.0
Clang version: Could not collect
CMake version: version 3.26.3
Libc version: N/A

Python version: 3.10.11 | packaged by Anaconda, Inc. | (main, Apr 20 2023, 18:56:50) [MSC v.1916 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.22631-SP0
Is CUDA available: True
CUDA runtime version: 12.1.105
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 4060 Laptop GPU
Nvidia driver version: 546.92
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture=9
CurrentClockSpeed=2200
DeviceID=CPU0
Family=207
L2CacheSize=32768
L2CacheSpeed=
Manufacturer=GenuineIntel
MaxClockSpeed=2200
Name=13th Gen Intel(R) Core(TM) i9-13900HX
ProcessorType=3
Revision=

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.25.2
[pip3] onnx==1.16.2
[pip3] onnxruntime==1.17.3
[pip3] onnxruntime-tools==1.7.0
[pip3] torch==2.0.1+cu118
[pip3] torchaudio==2.0.2+cu118
[pip3] torchdata==0.6.1
[pip3] torchinfo==1.8.0
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.15.2
[pip3] torchvision==0.15.2+cu118
[conda] blas                      1.0                         mkl    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] libblas                   3.9.0              12_win64_mkl    conda-forge
[conda] libcblas                  3.9.0              12_win64_mkl    conda-forge
[conda] liblapack                 3.9.0              12_win64_mkl    conda-forge
[conda] mkl                       2021.4.0           h0e2418a_729    conda-forge
[conda] mkl-service               2.4.0           py310hcf6e17e_0    conda-forge
[conda] numpy                     1.25.2                   pypi_0    pypi
[conda] torch                     2.0.1+cu118              pypi_0    pypi
[conda] torchaudio                2.0.2+cu118              pypi_0    pypi
[conda] torchdata                 0.6.1                    pypi_0    pypi
[conda] torchinfo                 1.8.0                    pypi_0    pypi
[conda] torchsummary              1.5.1                    pypi_0    pypi
[conda] torchtext                 0.15.2                   pypi_0    pypi
[conda] torchvision               0.15.2+cu118             pypi_0    pypi

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

**Comment by mingfeima** [comment]
@chunyuan-w please check what kind of onednn algo is used in the 2 scenarios above!

**Comment by NGC13009** [comment]
I just tried it out and when `dtype=torch.float64`  the `cpu` will also have no errors. When `torch.float32`  any `batch_size` used on the CPU will cause errors. 

```python
ci = 4              # channel in  : This problem only occurs when the number of input and output channels is greater than 1
co = 3              # channel out : This problem only occurs when the number of input and output channels is greater than 1
b = 2               # batch size, if change this, you need to synchronize and modify the following code.
h = 8               # feature's high
w = 6               # feature's wide
k = 3               # kernel size
device = "cpu"      # problem only occurs when use cpu
dtype = torch.float32

c2d = Conv2d(ci, co, k, 1, 0, 1, device=device, dtype=dtype)     # a simple net

for i in range(10):     # Multiple tests
    x = torch.randn(2, ci, h, w, device=device, dtype=dtype)
    y1 = c2d(x)                                                             # One batch in
    y2 = torch.cat([c2d(x[0:1, :, :, :]), c2d(x[1:2, :, :, :])], dim=0)     # One by one
    diff = torch.sum(torch.abs(y2 - y1), dim=[1, 2, 3])                     # the different sum of feature per batch
    print(diff)
```
the result is:
```text
tensor([3.5465e-06, 5.0913e-06], grad_fn=<SumBackward1>)
tensor([4.6017e-06, 4.5355e-06], grad_fn=<SumBackward1>)
tensor([3.7085e-06, 3.9315e-06], grad_fn=<SumBackward1>)
tensor([5.0266e-06, 4.3502e-06], grad_fn=<SumBackward1>)
tensor([4.0308e-06, 3.9656e-06], grad_fn=<SumBackward1>)
tensor([3.4748e-06, 3.9535e-06], grad_fn=<SumBackward1>)
tensor([4.3921e-06, 2.4401e-06], grad_fn=<SumBackward1>)
tensor([4.0866e-06, 4.1882e-06], grad_fn=<SumBackward1>)
tensor([3.3006e-06, 4.2040e-06], grad_fn=<SumBackward1>)
tensor([3.6065e-06, 4.2403e-06], grad_fn=<SumBackward1>)
```

**Comment by jgong5** [comment]
@NGC13009 Does the problem occur with multi-thread only or also happen on single thread?

**Comment by NGC13009** [comment]
> @NGC13009 Does the problem occur with multi-thread only or also happen on single thread?

My main program is single threaded startup. I'm not sure if multithreading was started during the computation.

**Comment by chunyuan-w** [comment]
> @chunyuan-w please check what kind of onednn algo is used in the 2 scenarios above!

Hi @mingfeima I checked that when `BS != 1`, it runs into oneDNN while when `BS == 1`, oneDNN is **NOT** chosen because the below condition is `false` thus this `use_mkldnn` function returns `false` in the reproducer when `BS == 1`:
https://github.com/pytorch/pytorch/blob/a00fad017719346bac6e08da0819358146e647e3/aten/src/ATen/native/Convolution.cpp#L530-L534

**Comment by mingfeima** [comment]
from @chunyuan-w 's comment, cpu chooses 2 different implementations when BS == 1 and BS != 1. 1e-6 is an acceptable tolerance on different algorithms, this is not a bug.
